File: src/main/scala/firelib/robot/ModelRuntimeContainer.cs
using System;
using System.IO;
using System.Linq;
using Fire.Common.Backtest;
using Fire.Common.Domain;
using Fire.Common.Threading;
using Fire.Common.Writers;
using NLog;
using QuantLib.Common;
using QuantLib.Domain;
using QuantLib.IndicatorBase;

namespace Fire.Common.Robot
{
    public class ModelRuntimeContainer
    {
        private readonly ModelRuntimeConfig modelRuntimeConfig;
        private readonly ThreadExecutor executor;
        private readonly IMarketDataProvider marketDataProvider;
        private readonly ITradeGate tradeGate;
        private readonly IModel model;
        private readonly MarketDataPlayer marketDataPlayer;
        private readonly Frequencer frequencer;
        private readonly IQuoteListener[] marketDataListeners;
        private readonly IStepListener[] stepListeners;
        private readonly RuntimeTradeWriter runtimeTradeWriter;

        private readonly Logger log;

        public ModelRuntimeContainer(ModelRuntimeConfig modelRuntimeConfig)
        {
            var threadName = modelRuntimeConfig.ModelConfig.ClassName;
           
            log = LogManager.GetLogger(threadName);

            log.Info("Starting config " + JsonSerializerUtil.ToStr(modelRuntimeConfig));
            this.modelRuntimeConfig = modelRuntimeConfig;

            runtimeTradeWriter = new RuntimeTradeWriter(Path.Combine(modelRuntimeConfig.TradeLogDirectory ?? Directory.GetCurrentDirectory(), "trades_live.csv"));

            executor = new ThreadExecutor(threadName: threadName);
            executor.Start();

            new ProvidersFactory().Create(modelRuntimeConfig, executor, out tradeGate, out marketDataProvider);

            Func<string, IMarketStub> marketStubFactory = (ti) => new MarketStubSwitcher(new MarketStub(ti), new ExecutionMarketStub(tradeGate, ti));

            var starter = new BacktesterSimple(marketStubFactory);

            MarketDataDistributor container;

            starter.RunSimple(modelRuntimeConfig.ModelConfig, out model, out marketDataPlayer, out container, modelRuntimeConfig.RunBacktest);
            marketDataListeners = marketDataPlayer.GetMarketDataListeners();
            stepListeners = marketDataPlayer.GetStepListeners();
            marketDataPlayer.Dispose();
            container.SwitchTickerType(TickerType.Tick);

            frequencer = new Frequencer(modelRuntimeConfig.ModelConfig.GetInterval());

            var stubSwitchers = model.GetStubs().Select(ms => (MarketStubSwitcher) ms);
            foreach (var switcher in stubSwitchers)
            {
                switcher.SwitchStubs();
                //to write trades to csv file
                switcher.AddCallback(new TradeGateCallbackAdapter((t) => runtimeTradeWriter.Write(modelRuntimeConfig.ModelConfig.ClassName, t)));
            }
            
            log.Info("Started ");
        }

        public void Start()
        {
            var interval = modelRuntimeConfig.ModelConfig.GetInterval();
            frequencer.AddListener((dt) => executor.Execute(() =>
                                                                {
                                                                    var roundTime = interval.RoundTime(DateTime.Now);
                                                                    foreach (var stepListener in stepListeners)
                                                                    {
                                                                        stepListener.OnStep(roundTime);
                                                                    }
                                                                }));


            foreach (var marketDataListener in marketDataListeners)
            {
                for (int k = 0; k < modelRuntimeConfig.ModelConfig.TickerIds.Count; k++)
                {
                    int k1 = k;
                    IQuoteListener listener = marketDataListener;
                    //!!!! assumed that market data provider works in the same thread
                    marketDataProvider.Subscribe(modelRuntimeConfig.ModelConfig.TickerIds[k].TickerId, (q) => listener.AddQuote(k1, q));
                }
            }


            frequencer.Start();
        }
    }
}

File: src/main/scala/firelib/robot/ModelRuntimeContainer.scala
package firelib.robot

import java.nio.file.Paths

import firelib.backtest._
import firelib.domain.Tick
import firelib.util.RuntimeTradeWriter
import org.joda.time.DateTime
import org.slf4j.LoggerFactory

class ModelRuntimeContainer(val modelRuntimeConfig: ModelRuntimeConfig) {
    var executor: ThreadExecutor;
    private var marketDataProvider: IMarketDataProvider
    private var tradeGate: ITradeGate
    private var model: IModel
    private var frequencer: Frequencer
    private val distributor: MarketDataDistributor

    private var stepListeners: Seq[IStepListener]


    val log = LoggerFactory.getLogger(getClass)


    log.info("Starting config " + JsonSerializerUtil.ToStr(modelRuntimeConfig));


    val tradesPath = Paths.get(modelRuntimeConfig.tradeLogDirectory.getOrElse("trades_live.csv")).toFile.getName


    executor = new ThreadExecutor(threadName = modelRuntimeConfig.modelConfig.ClassName);
    executor.Start();

    (tradeGate, marketDataProvider) = new ProvidersFactory().Create(modelRuntimeConfig, executor);

    val marketStubFactory = (ti: String) => new MarketStubSwitcher(new MarketStub(ti), new ExecutionMarketStub(tradeGate, ti));

    var starter = new BacktesterSimple(marketStubFactory);


    var marketDataPlayer: MarketDataPlayer
    //out
    (model, marketDataPlayer, distributor) = starter.RunSimple(modelRuntimeConfig.modelConfig, modelRuntimeConfig.RunBacktest)

    stepListeners = marketDataPlayer.getStepListeners();
    marketDataPlayer.Dispose();


    frequencer = new Frequencer(modelRuntimeConfig.modelConfig.interval);

    var stubSwitchers = model.stubs.map(ms => ms.asInstanceOf[MarketStubSwitcher])

    for (switcher <- stubSwitchers) {
        switcher.SwitchStubs();
        //to write trades to csv file
        switcher.AddCallback(new TradeGateCallbackAdapter((t) => RuntimeTradeWriter.write(tradesPath, modelRuntimeConfig.modelConfig.ClassName, t)));
    }

    log.info("Started ");


    def Start() = {
        var interval = modelRuntimeConfig.modelConfig.interval;

        frequencer.AddListener((dt) => {
            val act = () => stepListeners.foreach(_.OnStep(interval.RoundTime(DateTime.now())))
            executor.Execute(act)
        });


        for (k <- 0 until modelRuntimeConfig.modelConfig.TickerIds.length) {
            val k1 = k;
            //!!!! assumed that market data provider works in the same thread
            marketDataProvider.Subscribe(modelRuntimeConfig.modelConfig.TickerIds(k).TickerId, (q: Tick) => distributor.onTick(k1, q, null));
        }

        frequencer.Start();
    }
}
